rust-multiaddr/src/errors.rs

```diff
@@ -0,0 +1,71 @@
+use std::{net, fmt, error, io, num};
+use cid;
+use byteorder;
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+/// Error types
+#[derive(Debug)]
+pub enum Error {
+    UnknownProtocol,
+    UnknownProtocolString,
+    InvalidMultiaddr,
+    MissingAddress,
+    ParsingError(Box<error::Error + Send + Sync>),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(error::Error::description(self))
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::UnknownProtocol => "unknown protocol",
+            Error::UnknownProtocolString => "unknown protocol string",
+            Error::InvalidMultiaddr => "invalid multiaddr",
+            Error::MissingAddress => "protocol requires address, none given",
+            Error::ParsingError(_) => "failed to parse",
```

Comment:
Can we delegate to the inner error here?

Reply:
That's not the idiomatic thing to do, although I agree that the idiomatic thing to do is not really great.